# CSV og store datasett

In [3]:
inputfile = "EXR20250401.csv"
outputfile = "EXR-formattert.csv"

Eitt hovudmål i dette kurset er å kunna handtera store datasett frå røynda vha. programmering, i praksis i python. Me skal starta med datasettet, og so tek me programmeringsteknikkane etter kvart. Mange har gjort liknande ting i Excel eller andre rekneark, og då vil det ta ein del tid før føremonane ved programmering kjem til syne. Enkle oppgåver er som regel enklare å gjera, og i alle fall raskare å læra, i rekneark. Det er berre samansette oppgåver og store datasett som vert enklare ved programmering. Eitt problem med Excel spesifikt er maksgrensa på ein million radar. Me skal etter kvar sjå på datasett som er større enn det.

Merk at dette dokumentet er tenkt presentert munnleg i eit seminar.
Forklaringane kan difor vera litt knappe til sjølvstudium.
Det er likevel meininga at du skal fikla med koden min og gjera oppgåvene 
ved å føra inn nye celler i dokumentet.

Problemet me skal sjå på er valutakursar.  Korleis har dei utvikla seg dei siste fem åra?

## Fyrste datasett

Mange aktørar publiserer datasett. Me skal starta med valutakursar
[Noregs Bank](https://www.norges-bank.no/tema/Statistikk/Valutakurser/?tab=api).
Eg har lasta ned kursane for USD/EUR/GBP/DKK/SEK dei siste fem åra i fila 
[](EXR20250401.csv).

Dette er ei CSV-fil som står for «comma separated values». 
Namnet er ikkje alltid dekkjande; fila frå Noregs Bank bruker semikolon, og ikkje komma, som skiljeteikn.
Andre kjelder kan bruka tabulator eller røyr (`|`).
Dette må ein vera merksam på når ein skal lesa fila.

Fila er rein tekst, slik at me kan opna ho i ei teksteditor (vim, notepad, e.l.).
Dersom du vil opna ho i Excel, kan det løna seg å opna Excel med eitt tomt ark, og so bruka import-funksjonen frå menyane. Då skal du få opp ein dialogboks der du kan fortelja maskina at semikolon er skiljeteikn, og komma er det ikkje.

::: {admonition} Oppgåve
Opna fila i eit program du kjenner og sjå korleis ho ser ut.
Du kan godt opna både i ein teksteditor og i eit rekneark.
Kan du finna att kursen på pund sterling for fem år sidan?  Kva med svenske kroner 1. januar i år?
:::

Fila inneheld gjerne meir data enn me er interesserte i. Det skal me ikkje tenkja på.
Me legg vekt på å forstå det som er nyttig for oss.

::: {hint}
Når du skal arbeida med eit nytt datasett, løner det seg å sjå på fila for å forsikra deg om at du har fått det du ville.
:::
No kan me gå vidare til å opna fila i python.  Me bruker modulen pandas.

In [31]:
import pandas as pd
import datetime

df = pd.read_csv( inputfile, sep=";", decimal=",")

display(df)

,FREQ,Frekvens,BASE_CUR,Basisvaluta,QUOTE_CUR,Kvoteringsvaluta,TENOR,Løpetid,DECIMALS,CALCULATED,UNIT_MULT,Multiplikator,COLLECTION,Innsamlingstidspunkt,TIME_PERIOD,OBS_VALUE
0,B,Virkedag,DKK,Danske kroner,NOK,Norske kroner,SP,Spot,2,False,2,Hundre,C,ECB concertation tidspunkt 14:15 CET,2020-04-01,150.97
1,B,Virkedag,DKK,Danske kroner,NOK,Norske kroner,SP,Spot,2,False,2,Hundre,C,ECB concertation tidspunkt 14:15 CET,2020-04-02,150.47
2,B,Virkedag,DKK,Danske kroner,NOK,Norske kroner,SP,Spot,2,False,2,Hundre,C,ECB concertation tidspunkt 14:15 CET,2020-04-03,150.80
3,B,Virkedag,DKK,Danske kroner,NOK,Norske kroner,SP,Spot,2,False,2,Hundre,C,ECB concertation tidspunkt 14:15 CET,2020-04-06,152.52
4,B,Virkedag,DKK,Danske kroner,NOK,Norske kroner,SP,Spot,2,False,2,Hundre,C,ECB concertation tidspunkt 14:15 CET,2020-04-07,148.57
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
6295,B,Virkedag,SEK,Svenske kroner,NOK,Norske kroner,SP,Spot,2,False,2,Hundre,C,ECB concertation tidspunkt 14:15 CET,2025-03-25,104.74
6296,B,Virkedag,SEK,Svenske kroner,NOK,Norske kroner,SP,Spot,2,False,2,Hundre,C,ECB concertation tidspunkt 14:15 CET,2025-03-26,104.79
6297,B,Virkedag,SEK,Svenske kroner,NOK,Norske kroner,SP,Spot,2,False,2,Hundre,C,ECB concertation tidspunkt 14:15 CET,2025-03-27,104.80
6298,B,Virkedag,SEK,Svenske kroner,NOK,Norske kroner,SP,Spot,2,False,2,Hundre,C,ECB concertation tidspunkt 14:15 CET,2025-03-28,104.38


In [37]:
df["TIME_PERIOD"] = pd.PeriodIndex( df["TIME_PERIOD"],freq="D" )
display( df )
df.dtypes

,FREQ,Frekvens,BASE_CUR,Basisvaluta,QUOTE_CUR,Kvoteringsvaluta,TENOR,Løpetid,DECIMALS,CALCULATED,UNIT_MULT,Multiplikator,COLLECTION,Innsamlingstidspunkt,TIME_PERIOD,OBS_VALUE
0,B,Virkedag,DKK,Danske kroner,NOK,Norske kroner,SP,Spot,2,False,2,Hundre,C,ECB concertation tidspunkt 14:15 CET,2020-04-01,150.97
1,B,Virkedag,DKK,Danske kroner,NOK,Norske kroner,SP,Spot,2,False,2,Hundre,C,ECB concertation tidspunkt 14:15 CET,2020-04-02,150.47
2,B,Virkedag,DKK,Danske kroner,NOK,Norske kroner,SP,Spot,2,False,2,Hundre,C,ECB concertation tidspunkt 14:15 CET,2020-04-03,150.80
3,B,Virkedag,DKK,Danske kroner,NOK,Norske kroner,SP,Spot,2,False,2,Hundre,C,ECB concertation tidspunkt 14:15 CET,2020-04-06,152.52
4,B,Virkedag,DKK,Danske kroner,NOK,Norske kroner,SP,Spot,2,False,2,Hundre,C,ECB concertation tidspunkt 14:15 CET,2020-04-07,148.57
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
6295,B,Virkedag,SEK,Svenske kroner,NOK,Norske kroner,SP,Spot,2,False,2,Hundre,C,ECB concertation tidspunkt 14:15 CET,2025-03-25,104.74
6296,B,Virkedag,SEK,Svenske kroner,NOK,Norske kroner,SP,Spot,2,False,2,Hundre,C,ECB concertation tidspunkt 14:15 CET,2025-03-26,104.79
6297,B,Virkedag,SEK,Svenske kroner,NOK,Norske kroner,SP,Spot,2,False,2,Hundre,C,ECB concertation tidspunkt 14:15 CET,2025-03-27,104.80
6298,B,Virkedag,SEK,Svenske kroner,NOK,Norske kroner,SP,Spot,2,False,2,Hundre,C,ECB concertation tidspunkt 14:15 CET,2025-03-28,104.38


FREQ                          str
Frekvens                      str
BASE_CUR                      str
Basisvaluta                   str
QUOTE_CUR                     str
Kvoteringsvaluta              str
TENOR                         str
Løpetid                       str
DECIMALS                    int64
CALCULATED                   bool
UNIT_MULT                   int64
Multiplikator                 str
COLLECTION                    str
Innsamlingstidspunkt          str
TIME_PERIOD             period[D]
OBS_VALUE                 float64
dtype: object

In [38]:
df1 = df[ ["BASE_CUR", "TIME_PERIOD", "OBS_VALUE" ] ]
df1

,BASE_CUR,TIME_PERIOD,OBS_VALUE
0,DKK,2020-04-01,150.97
1,DKK,2020-04-02,150.47
2,DKK,2020-04-03,150.80
3,DKK,2020-04-06,152.52
4,DKK,2020-04-07,148.57
...,...,...,...
6295,SEK,2025-03-25,104.74
6296,SEK,2025-03-26,104.79
6297,SEK,2025-03-27,104.80
6298,SEK,2025-03-28,104.38


In [39]:
valutakurs = df1.pivot(columns="BASE_CUR", index="TIME_PERIOD" )
display(valutakurs)

OBS_VALUE                                   
BASE_CUR          DKK      EUR      GBP     SEK      USD
TIME_PERIOD                                             
2020-04-01     150.97  11.2685  12.7385  103.03  10.3040
2020-04-02     150.47  11.2345  12.8046  102.82  10.3012
2020-04-03     150.80  11.2628  12.8205  102.84  10.4430
2020-04-06     152.52  11.3875  12.9698  103.72  10.5528
2020-04-07     148.57  11.0945  12.5941  101.98  10.1925
...               ...      ...      ...     ...      ...
2025-03-25     151.60  11.3090  13.5332  104.74  10.4471
2025-03-26     152.25  11.3585  13.5830  104.79  10.5288
2025-03-27     152.05  11.3435  13.6147  104.80  10.5178
2025-03-28     151.36  11.2940  13.5488  104.38  10.4603
2025-03-31     152.96  11.4130  13.6624  105.20  10.5529

[1260 rows x 5 columns]

In [45]:

display(valutakurs.transpose())

TIME_PERIOD         2020-04-01  2020-04-02  2020-04-03  2020-04-06  \
          BASE_CUR                                                   
OBS_VALUE DKK         150.9700    150.4700    150.8000    152.5200   
          EUR          11.2685     11.2345     11.2628     11.3875   
          GBP          12.7385     12.8046     12.8205     12.9698   
          SEK         103.0300    102.8200    102.8400    103.7200   
          USD          10.3040     10.3012     10.4430     10.5528   

TIME_PERIOD         2020-04-07  2020-04-08  2020-04-14  2020-04-15  \
          BASE_CUR                                                   
OBS_VALUE DKK         148.5700    149.9700    151.5900    153.4100   
          EUR          11.0945     11.1940     11.3118     11.4503   
          GBP          12.5941     12.7280     12.9644     13.1033   
          SEK         101.9800    102.3400    103.6300    104.7400   
          USD          10.1925     10.2971     10.3182     10.5020   

TIME_PERIOD         2020-04-16  2020-04-17  ...  2025-03-18  2025-03-19  \
          BASE_CUR                          ...                           
OBS_VALUE DKK         152.9700    151.2600  ...    154.5300    154.2400   
          EUR          11.4145     11.2845  ...     11.5280     11.5050   
          GBP          13.0956     12.9740  ...     13.6956     13.6837   
          SEK         104.8200    103.9600  ...    104.8300    104.5700   
          USD          10.4836     10.3909  ...     10.5587     10.5580   

TIME_PERIOD         2025-03-20  2025-03-21  2025-03-24  2025-03-25  \
          BASE_CUR                                                   
OBS_VALUE DKK         154.2000    152.8400    152.5600    151.6000   
          EUR          11.5025     11.4010     11.3795     11.3090   
          GBP          13.7442     13.6107     13.6016     13.5332   
          SEK         104.2900    104.2800    103.8600    104.7400   
          USD          10.6180     10.5302     10.5132     10.4471   

TIME_PERIOD         2025-03-26  2025-03-27  2025-03-28  2025-03-31  
          BASE_CUR                                                  
OBS_VALUE DKK         152.2500    152.0500    151.3600    152.9600  
          EUR          11.3585     11.3435     11.2940     11.4130  
          GBP          13.5830     13.6147     13.5488     13.6624  
          SEK         104.7900    104.8000    104.3800    105.2000  
          USD          10.5288     10.5178     10.4603     10.5529  

[5 rows x 1260 columns]

In [41]:
valutakurs.index

PeriodIndex(['2020-04-01', '2020-04-02', '2020-04-03', '2020-04-06',
             '2020-04-07', '2020-04-08', '2020-04-14', '2020-04-15',
             '2020-04-16', '2020-04-17',
             ...
             '2025-03-18', '2025-03-19', '2025-03-20', '2025-03-21',
             '2025-03-24', '2025-03-25', '2025-03-26', '2025-03-27',
             '2025-03-28', '2025-03-31'],
            dtype='period[D]', name='TIME_PERIOD', length=1260)

In [42]:
valutakurs.dtypes

           BASE_CUR
OBS_VALUE  DKK         float64
           EUR         float64
           GBP         float64
           SEK         float64
           USD         float64
dtype: object

In [43]:
display( valutakurs )

OBS_VALUE                                   
BASE_CUR          DKK      EUR      GBP     SEK      USD
TIME_PERIOD                                             
2020-04-01     150.97  11.2685  12.7385  103.03  10.3040
2020-04-02     150.47  11.2345  12.8046  102.82  10.3012
2020-04-03     150.80  11.2628  12.8205  102.84  10.4430
2020-04-06     152.52  11.3875  12.9698  103.72  10.5528
2020-04-07     148.57  11.0945  12.5941  101.98  10.1925
...               ...      ...      ...     ...      ...
2025-03-25     151.60  11.3090  13.5332  104.74  10.4471
2025-03-26     152.25  11.3585  13.5830  104.79  10.5288
2025-03-27     152.05  11.3435  13.6147  104.80  10.5178
2025-03-28     151.36  11.2940  13.5488  104.38  10.4603
2025-03-31     152.96  11.4130  13.6624  105.20  10.5529

[1260 rows x 5 columns]

In [44]:
valutakurs.transpose().iloc[1]

TIME_PERIOD
2020-04-01    11.2685
2020-04-02    11.2345
2020-04-03    11.2628
2020-04-06    11.3875
2020-04-07    11.0945
               ...   
2025-03-25    11.3090
2025-03-26    11.3585
2025-03-27    11.3435
2025-03-28    11.2940
2025-03-31    11.4130
Freq: D, Name: (OBS_VALUE, EUR), Length: 1260, dtype: float64

**Merk**
1. Me måtte spesifisera semikolon som skiljeteikn (`sep=";"`).
2. Berre `df` (evt. `display(df)`, og ikkje `print(df)` gjev den penaste utskrifta.
3. Me ser dei fyrste og dei siste linene i fila.
4. Me har fått ei ekstra indekssøyle til venstre, som ikkje finst i fila. Dette er berre for å nummerera radane.

Neste steg er å visualisera valutakursane.  Litt naïvt kan me freista `.plot()`:

In [46]:
t = valutakurs.transpose()
display(t)

TIME_PERIOD         2020-04-01  2020-04-02  2020-04-03  2020-04-06  \
          BASE_CUR                                                   
OBS_VALUE DKK         150.9700    150.4700    150.8000    152.5200   
          EUR          11.2685     11.2345     11.2628     11.3875   
          GBP          12.7385     12.8046     12.8205     12.9698   
          SEK         103.0300    102.8200    102.8400    103.7200   
          USD          10.3040     10.3012     10.4430     10.5528   

TIME_PERIOD         2020-04-07  2020-04-08  2020-04-14  2020-04-15  \
          BASE_CUR                                                   
OBS_VALUE DKK         148.5700    149.9700    151.5900    153.4100   
          EUR          11.0945     11.1940     11.3118     11.4503   
          GBP          12.5941     12.7280     12.9644     13.1033   
          SEK         101.9800    102.3400    103.6300    104.7400   
          USD          10.1925     10.2971     10.3182     10.5020   

TIME_PERIOD         2020-04-16  2020-04-17  ...  2025-03-18  2025-03-19  \
          BASE_CUR                          ...                           
OBS_VALUE DKK         152.9700    151.2600  ...    154.5300    154.2400   
          EUR          11.4145     11.2845  ...     11.5280     11.5050   
          GBP          13.0956     12.9740  ...     13.6956     13.6837   
          SEK         104.8200    103.9600  ...    104.8300    104.5700   
          USD          10.4836     10.3909  ...     10.5587     10.5580   

TIME_PERIOD         2025-03-20  2025-03-21  2025-03-24  2025-03-25  \
          BASE_CUR                                                   
OBS_VALUE DKK         154.2000    152.8400    152.5600    151.6000   
          EUR          11.5025     11.4010     11.3795     11.3090   
          GBP          13.7442     13.6107     13.6016     13.5332   
          SEK         104.2900    104.2800    103.8600    104.7400   
          USD          10.6180     10.5302     10.5132     10.4471   

TIME_PERIOD         2025-03-26  2025-03-27  2025-03-28  2025-03-31  
          BASE_CUR                                                  
OBS_VALUE DKK         152.2500    152.0500    151.3600    152.9600  
          EUR          11.3585     11.3435     11.2940     11.4130  
          GBP          13.5830     13.6147     13.5488     13.6624  
          SEK         104.7900    104.8000    104.3800    105.2000  
          USD          10.5288     10.5178     10.4603     10.5529  

[5 rows x 1260 columns]

In [61]:
t.loc[("OBS_VALUE","DKK"),pd.Period("2020-04-01",freq="D")] = 11


In [62]:
display(t)


TIME_PERIOD         2020-04-01  2020-04-02  2020-04-03  2020-04-06  \
          BASE_CUR                                                   
OBS_VALUE DKK          11.0000    150.4700    150.8000    152.5200   
          EUR          11.2685     11.2345     11.2628     11.3875   
          GBP          12.7385     12.8046     12.8205     12.9698   
          SEK         103.0300    102.8200    102.8400    103.7200   
          USD          10.3040     10.3012     10.4430     10.5528   
0                      11.0000         NaN         NaN         NaN   

TIME_PERIOD         2020-04-07  2020-04-08  2020-04-14  2020-04-15  \
          BASE_CUR                                                   
OBS_VALUE DKK         148.5700    149.9700    151.5900    153.4100   
          EUR          11.0945     11.1940     11.3118     11.4503   
          GBP          12.5941     12.7280     12.9644     13.1033   
          SEK         101.9800    102.3400    103.6300    104.7400   
          USD          10.1925     10.2971     10.3182     10.5020   
0                          NaN         NaN         NaN         NaN   

TIME_PERIOD         2020-04-16  2020-04-17  ...  2025-03-18  2025-03-19  \
          BASE_CUR                          ...                           
OBS_VALUE DKK         152.9700    151.2600  ...    154.5300    154.2400   
          EUR          11.4145     11.2845  ...     11.5280     11.5050   
          GBP          13.0956     12.9740  ...     13.6956     13.6837   
          SEK         104.8200    103.9600  ...    104.8300    104.5700   
          USD          10.4836     10.3909  ...     10.5587     10.5580   
0                          NaN         NaN  ...         NaN         NaN   

TIME_PERIOD         2025-03-20  2025-03-21  2025-03-24  2025-03-25  \
          BASE_CUR                                                   
OBS_VALUE DKK         154.2000    152.8400    152.5600    151.6000   
          EUR          11.5025     11.4010     11.3795     11.3090   
          GBP          13.7442     13.6107     13.6016     13.5332   
          SEK         104.2900    104.2800    103.8600    104.7400   
          USD          10.6180     10.5302     10.5132     10.4471   
0                          NaN         NaN         NaN         NaN   

TIME_PERIOD         2025-03-26  2025-03-27  2025-03-28  2025-03-31  
          BASE_CUR                                                  
OBS_VALUE DKK         152.2500    152.0500    151.3600    152.9600  
          EUR          11.3585     11.3435     11.2940     11.4130  
          GBP          13.5830     13.6147     13.5488     13.6624  
          SEK         104.7900    104.8000    104.3800    105.2000  
          USD          10.5288     10.5178     10.4603     10.5529  
0                          NaN         NaN         NaN         NaN  

[6 rows x 1260 columns]

In [63]:
display(df)

,FREQ,Frekvens,BASE_CUR,Basisvaluta,QUOTE_CUR,Kvoteringsvaluta,TENOR,Løpetid,DECIMALS,CALCULATED,UNIT_MULT,Multiplikator,COLLECTION,Innsamlingstidspunkt,TIME_PERIOD,OBS_VALUE
0,B,Virkedag,DKK,Danske kroner,NOK,Norske kroner,SP,Spot,2,False,2,Hundre,C,ECB concertation tidspunkt 14:15 CET,2020-04-01,150.97
1,B,Virkedag,DKK,Danske kroner,NOK,Norske kroner,SP,Spot,2,False,2,Hundre,C,ECB concertation tidspunkt 14:15 CET,2020-04-02,150.47
2,B,Virkedag,DKK,Danske kroner,NOK,Norske kroner,SP,Spot,2,False,2,Hundre,C,ECB concertation tidspunkt 14:15 CET,2020-04-03,150.80
3,B,Virkedag,DKK,Danske kroner,NOK,Norske kroner,SP,Spot,2,False,2,Hundre,C,ECB concertation tidspunkt 14:15 CET,2020-04-06,152.52
4,B,Virkedag,DKK,Danske kroner,NOK,Norske kroner,SP,Spot,2,False,2,Hundre,C,ECB concertation tidspunkt 14:15 CET,2020-04-07,148.57
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
6295,B,Virkedag,SEK,Svenske kroner,NOK,Norske kroner,SP,Spot,2,False,2,Hundre,C,ECB concertation tidspunkt 14:15 CET,2025-03-25,104.74
6296,B,Virkedag,SEK,Svenske kroner,NOK,Norske kroner,SP,Spot,2,False,2,Hundre,C,ECB concertation tidspunkt 14:15 CET,2025-03-26,104.79
6297,B,Virkedag,SEK,Svenske kroner,NOK,Norske kroner,SP,Spot,2,False,2,Hundre,C,ECB concertation tidspunkt 14:15 CET,2025-03-27,104.80
6298,B,Virkedag,SEK,Svenske kroner,NOK,Norske kroner,SP,Spot,2,False,2,Hundre,C,ECB concertation tidspunkt 14:15 CET,2025-03-28,104.38


In [ ]:
df.plot()

OK.  Me kan i alle fall sjå at `plot()` finst som funksjon og faktisk teiknar eit plott, sjølv om dette plottet ikkje er nyttig. Me har jo ikkje sagt kva søyler me vil ha plotta.

## Ein liten digresjon om funksjonsnotasjon

Lat oss dvela litt ved notasjonen. Me har sett fleire variantar av funksjonar.
1. `print( df )` 
2. `pd.read_csv("EXR20250401.csv", sep=";")`
3. `df.plot()`

Kvifor heiter det ikkje `df.print()` eller `plot(df)`?

Grunnen er kvar funksjonen er definert.  I det fyrste dømet er `print` ein *global* funksjon.
Han er alltid tilgjengeleg.  Argumentet `df` fortel kva data `print` skal prenta.

I det andre dømet hentar me funksjonen `read_csv` frå ein modul, pandas, som me døypte `pd` då me importerte det. Då er `pd` kjelda der me finn `read_csv` og `"EXR20250401.csv", sep=";"` er filnamn og konfigurasjon som funksjonen opererer på.

I det tredje dømet er funksjonen `plot()` definert i *objektet* `df`, som fekk verdien sin frå `pd.read_csv(...)` tidlegare.  Her treng me ikkje noko argument, fordi funksjonen allereie kjem frå, og veit om, `df`, og freistar å plotta dette objektet.


## Plottet

::: {admonition} Refleksjon
Kva søyler er interessante å plotta?
:::

Me kan freista med `TIME_PERIOD` og `OBS_VALUE` som ser ut som dato og kurs.

```ipython3
df.plot( x="TIME_PERIOD", y="OBS_VALUE" )
```

::: {admonition} Oppgåve
Lag ein kodeblokk og test kodelina over.
:::

Dette er òg for naivt.
Om du set denne koden i ein kode-blokk, får du *«No numeric data to plot»*.
Det er kanskje litt overraskande, men lat oss sjå nærare på dei to søylene.
Me kan be pandas om datatypane for alle søylene, slik:

In [ ]:
df.dtypes

Dette var rart.  Der er to numeriske søyler, `DECIMALS` og `UNIT_MULT`. Dei er heiltal i intervallet $-2^{63}\ldots2^{63}-1$ (`int64`), men dei varierer lite og er neppe interessante.  På den andre sida, kan me no sjå at dette var dei to søylene som fyrst vart plotta.

Dei to søylene som interesserer oss har typen `object`, noko som kan vera nær sagt kva som helst, men ikkje noko som pandas veit å plotta. Me må altso læra å omsetja desse søylene til noko numerisk. 

## Datatypar

Datattypar kan skapa mykje forvirring når me arbeider med data, og det kan vera komplisert fordi der er uendeleg mange datatypar, og kvar type vert representert ulikt internt i maskinen og overfor brukaren. Internt i maskina er alt representert som ein serie med *bits*, der kvar *bit* er 0 eller 1, av eller på, sann eller usann; kjært barn har mange namn.  Har ein åtte *bits* (ein *byte*), kan ein tolka dei saman som eit heiltal mellom 0 og 255, eller som eit teikn (bokstav eller anna) frå ein eller annan tabell 
(t.d. [ASCII](https://en.wikipedia.org/wiki/ASCII)).
Har ein fleire *bytes* kan ein tolka det som ein teiknstreng (t.d. `"Hello World"`).

I python har me sett at teiknstrengar vert skrivne i gåseaugo, medan tal vert skrivne utan.  Difor er `100` ikkje det same som `"100"` eller `"eitt hundrede"`.  For oss kan det vera det same, men for maskina er det berre `100` som er eit tal.

I CSV-fila er alt teiknstrengar.  Det er gjort slik for at me skal kunna opna og manipulera filen enklast mogleg med ulike verkty og program. Det som framstår som `object` over er teiknstrengar som pandas ikkje klarte å forstå som noko anna. Dei andre to typane er

+ `bool`, som er kort for *Boolean* eller bolsk på norsk.  Denne typen har verdiane sann eller usann (`True` eller `False`)
+ `int64` er kort for *integer* eller heiltal, med 64 bits, som er nok for å representera positive og negative tal opp til $2^{63}$ eller cirka $8\cdot10^{18}$. Dersom me prøver å lagra tal som er større enn dette, får me ein overflytfeil. Nokre system kan varsla om slike feil, men ofte kan slike feil vera vanskeleg å finna.

Me skal straks sjå to typar til

+ `float64` er desimaltal med 64 *bits*.  *Float* er kort for *floating point number*, eller flyttal på norsk. Dette vil seia at desimalkommaet «flyt», slik at ein bruker mange desimalplassar for små tal, og få for store. Flyttal er i dag einerådande som representasjon av desimaltal. *Fixed point numbers* finst i ein del eldre programmeringsspråk, men då skal me tilbake til 1970- og 60-talet.
+ `datetime64` representerer eit tidspunkt, her ogso med 64 bits.

Det er typisk å bruka 64-*bits* tal fordi prosessorane (CPU) i dag stort sett er 64-*bits* og dermed handsamar 64-*bits*-verdiar som ei eining.  Bruker ein fleire *bits* må prosessoren handsame det i fleire omgongar, og bruker ein færre «sløser» ein med prosessorkapasiteten.


## Problemet frå Noregs Bank

**Kursen** Denne søyla ser ut som desimaltal, og det kan verka rart at pandas ikkje skjøner det. Ser me litt nøyare etter ser me at datafila bruker desimal*komma* medan pandas ventar seg det engelske desimal*punktumet*. Det er altso dette me må fiksa. Me treng då to steg, fyrst må me byta ut desimalkomma i heile søyla, og deretter må me be pandas tolka ho på nytt som tal, t.d. slik:

In [ ]:
df['engelsk'] = df['OBS_VALUE'].str.replace(',', '.')
df['kurs'] = pd.to_numeric(df['engelsk'])
df.dtypes

Merk notasjonen.  Klammene, i `df['OBS_VALUE']` let oss adressera éi søyle i datasettet. Dette kan me gjera både på venstre og høgre side av tilordninga (`=`).  På venstre side tilordnar me ei ny søyle («engelsk» og «kurs»). På høgre side bruker me ei eksisterande søyle i utrekninga til den nye. Slik me har gjort det her, vert den gamle søyla (`OBS_VALUE`) ikkje endra.

Den fyrste lina bruker strengfunksjonen `replace` på kvar einaste verdi i søyla. Den andre lina bruker konverteringsrutinen `to_numeric` for å lesa eit tal frå teiknstrengen. I utlistinga ser me at datasettet har fått to nye søyler, der `"kurs"` er flyttal.  Lat oss sjå på heile datasettet.

In [ ]:
display(df)

Dette ser jo bra ut, men so må me gjera noko med datosøyla.  Dette er meir komplisert fordi der er so frykteleg mange måtar å skriva tidspunkt på.  Me skal bruka [`to_datetime`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.to_datetime.html)-funksjonen, som fungerer litt som `to_numeric`, bortsett frå at me må fortelja han korleis datoen er formatter, slik:

In [ ]:
df['dato'] = pd.to_datetime(df['TIME_PERIOD'], format='%Y-%m-%d')
df.dtypes

In [ ]:
display(df)

Her har eg laga nye søyler. Det er vanleg å overskriva eksisterande søyler når ein formatterer dei, t.d. slik
```
df['OBS_VALUE'] = pd.to_numeric( df['OBS_VALUE'].str.replace(',', '.') )
df['TIME_PERIOD'] = pd.to_datetime(df['TIME_PERIOD'], format='%Y-%m-%d')
```
Grunnen til at eg ikkje har gjort det her, er dels for å kunne samanlikna den nye og den gamle søyla, og dels fordi eg då ikkje risikerer å bruka feil søyle om eg skulle gå tilbake i *notebook*-fila og køyra ei gamal celle omatt.


## Tilbake til plotting

In [ ]:
df.plot( "dato", "kurs")

Dette så rart ut, men lat oss ta det positive fyrst.
1. $x$-aksen ser ut som dato dei siste fem åra
2. $y$-aksen har talverdiar som kan vera valutakursar, mellom 10 og 150.
3. Me har kurvar som går litt opp og ned som valutakursar gjer.

Det er ein god start, men kva er dei rake krosslinene?

Problemet finn me i fila, som har fem ulike valutaar. Dei fem valuataane har ikkje kvar si søyle.
I staden er der ei søyle for valutakode.

::: {admonition} Refleksjonsspørsmål
Kva del av plottet svarer til kva valuta?
:::

For å få eit godt plott, må me filtrera datasettet for å skilja dei fem valutaane frå kvarandre.

In [ ]:
gbp = df[ df["BASE_CUR"] == "GBP" ]
display( gbp )
gbp.plot( "dato", "kurs" )

Det såg betre ut. Me kan freista å gjera datasettet litt ryddigare ved å filtrera ut søyler og byta namn på dei.

In [ ]:
gbp = gbp.reset_index()
gbp["GBP"] = gbp["kurs"]
gbp = gbp.filter( items= [ "dato", "GBP" ] )
gbp

Merk at `reset_index()` gjer to ting. For det fyrste renummerer han radane, slik at dei startar på null. For det andre vert datasettet kopiert. Opprinneleg viste `gbp` berre til eit stykke (*slice*) av `df`, og då ville me ikkje hatt lov til å laga nye søyler slik som me gjer i den andre lina.

Denne omnamninga er ein god start til å plotta fleire kursar i eitt diagram.
Fyrst lagar me eit datasett for danske kroner.

In [ ]:
dkk = df[ df["BASE_CUR"] == "DKK" ]
dkk = dkk.reset_index()
dkk["DKK"] = dkk["kurs"]
dkk = dkk.filter( items= [ "dato", "DKK" ] )
dkk

So kan me fletta dei to datasetta saman.  Me vil fletta på datosøyla. 
Sannsynlegvis speler ikkje det noka rolle her, sidan me reknar med at båe valutaane er notert på alle dei same datoane.
Generelt er det greitt å tenkja på, sidan somme verdiar ofte manglar i faktiske datasett.
For å vera på den sikre sida, bruker me `on`-parameteren. Då får me ei line for kvar dato som finst i eitt av datasetta.

In [ ]:
mrg = pd.merge_asof( dkk, gbp, on="dato" )
display(mrg)
mrg.plot()

OK.  Datasettet ser fint ut, men plottet var ikkje det som me venta.

::: {admonition} Refleksjon
Kva er det eigentleg som er plotta?
:::

Når me vel $x$- og $y$-akse i `plot`-funksjonen, kan me gje ei liste med søyler, slik:

In [ ]:
mrg.plot( x="dato", y = [ "GBP", "DKK" ])

::: {admonition} Oppgåve  
Bruk det som me har lært til no for å plotta alle dei fem valutaane i eitt diagram.
:::


## Skriva ut ein fil

Heilt til slutt skal me sjå korleis me kan skriva det reformatterte datasettet tilbake til fil.

In [ ]:
# Skriva ut CSV-fila
mrg.to_csv(outputfile, index=False)

Me ser ikkje noko utdata frå denne funksjonen, so for å sjå kva som skjedde, må me finna att fila i filsystemet.

::: {admonition} Oppgåve
Finn fila i filsystemet og opna ho i Excel eller eit anna program. Inneheld ho det ho skal?
:::

## Avrunding

Der er mykje å halda styr på i pandas. Her har me berre sett eitt døme og nokre få teknikkar. Når du skal bruka python og pandas i praksis vil du stadig vekk ha bruk for å slå opp konkrete teknikkar.  T.d.

* [Cheat sheet](https://pandas.pydata.org/Pandas_Cheat_Sheet.pdf)
* [Tutorials](https://pandas.pydata.org/pandas-docs/stable/getting_started/tutorials.html)
* [Kokebok](https://www.skytowner.com/explore/pandas_recipes_reference) 

Det er ikkje sikkert at det er mykje hjelp i desse kjeldene enno, før du har lært litt meir. I neste demo skal me forklara nokre fleire grunnleggjande konsept. Målet er å læra nok til å søkja etter resten, lesa dokumentasjon og å stilla gode spørsmål når du treng det.

::: {admonition} Oppgåve
Kva kan du seia om valutakursane basert på dei plotta og visningane som me har laga?
:::